pyshacl/cli.py

```diff
@@ -183,16 +185,13 @@ def str_is_true(s_var: str):
 # parser.add_argument('-h', '--help', action="help", help='Show this help text.')
 
 
-def main():
+def main() -> None:
```

Comment:
Should `main()` be changed to return `int` exit code, then pass that to `sys.exit()` from the caller site? That would avoid all the `sys.exit()` calls in here.

Reply:
At some point years ago, the idea wormed into my head that `main()` in Python had to return `None`.

However, I've had poor luck finding out where that idea came from.  My current best guess is `mypy` complaining about missing return signatures when I was first learning type annotations.  E.g. from this program `try.py`:

```python
def main():
    x = 1
    x = "1"

if __name__ == "__main__":
    main()
```

`mypy try.py` returns no issues; fine & dandy, there's no typed context.  `mypy --strict` catches the error I want around `x`, but gives this output about the rest:

```
try.py:1: error: Function is missing a return type annotation  [no-untyped-def]
try.py:1: note: Use "-> None" if function does not return a value
try.py:3: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
try.py:6: error: Call to untyped function "main" in typed context  [no-untyped-call]
Found 3 errors in 1 file (checked 1 source file)
```

I thought there used to be a specific suggestion about `main` instead of what's on the first 2 or last 1 lines.  It might have just been me conflating nearby lines instead.  I'm not finding a say on the matter in `mypy`'s documentation, PEPs on typing (483 and 484), or other search engine results.

Since `mypy --strict` didn't complain at this:

```python
from typing import Dict, Set

def main() -> Dict[str, Set[int]]:
    return {"a": {1}}

if __name__ == "__main__":
    main()
```

I'm fine swapping `-> None` with anything else if you think it'll result in cleaner chaining.

Reply:
I started the "`main` returns an `int`" revision, and stopped partway through revising the `sys.exit(2)` calls when I remembered another hint I had for why `main` usually returns `None`.

`sys.exit(n)` influences the return status of the program call from `main`.  `return n` doesn't.

E.g. here's a file `try.py`:

```python
def main() -> int:
    print("I ran.")
    return 1
main()
```

Here's the shell-visible exit status:

```
$ python3 try.py ; echo $?
I ran.
0
```

So I think the more shell-consistent idiom is to have the calls to `sys.exit(main())` instead just be calls to `main()`, and to leave any deeper-in-`main` calls to `sys.exit(n)` as they are now.  I don't think having `main() -> int` gains any control-flow advantage over `sys.exit` doing hard-stops with a controlled exit status value, or `raise SomethingUnfortunate` passing a traceback exit value 1 if uncaught.